### PR TITLE
spec: add os/unix/sysctl isolator

### LINKED ISSUE #645: ace: add sysctl isolator
Applications running inside pods may need to tweak several namespaced sysctl parameters. There is a need for executor to support this (https://github.com/coreos/rkt/issues/2694), and OCI Linux runtime manifest already features a sysctl stanza (https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#sysctl).

In short, ACE should provide an `os/linux/sysctl` isolator.

### PR DESCRIPTION
**Pull request by lucab**
This PR introduces an "os/unix/sysctl" isolator, which applies at pod-level.

Closes #645

---

Arbitrary/controversial choices made here:
 * isolator type, as it applies inside pod-context (like other isolators) instead of the host (like top-level stanzas: volumes, ports)
 * ~~new "environment" isolator category, as it can also accommodate "ulimit" ones~~

### CODE REVIEW

File: spec/ace.md [hunk 1]
```diff
@@ -383,6 +383,33 @@ Small quantities can be represented directly as decimals (e.g., 0.3), or using m
 
 **NOTE**: Network limits MUST NOT apply to localhost communication between apps in a pod.
 
+### Environment Isolators
+
+An application or a group of applications may often require some tweaking to the environment where they run, in order to perform in an optimal way.
```
Comment (euank) [inline spec/ace.md]: Awkward wording I think. I'll suggest "An application may want or need some changes to the system environment it is run in."

At the least, nix often.

File: spec/ace.md [hunk 2]
```diff
@@ -383,6 +383,33 @@ Small quantities can be represented directly as decimals (e.g., 0.3), or using m
...
+Environment isolators take care of setting up the runtime environment as needed.
+
+#### environment/sysctl
+
```
Comment (euank) [inline spec/ace.md]: Paragraph explaining what sysctls roughly are and that some are namespaced / attached to certain namespaces.

File: spec/ace.md [hunk 3]
```diff
@@ -383,6 +383,33 @@ Small quantities can be represented directly as decimals (e.g., 0.3), or using m
...
+  "sysctl": {
+    "net.ipv4.tcp_mtu_probing": "1",
+    "net.ipv4.tcp_keepalive_time": "300",
+    "net.ipv4.tcp_keepalive_probes": "5",
+    "net.ipv4.tcp_keepalive_intvl": "15",
+  }
+}
+```
+**Notes**:
+ 1. Only a single `environment/sysctl` isolator can be specified per-pod.
```
Comment (euank) [inline spec/ace.md]: Why not merge?

File: spec/ace.md [hunk 4]
```diff
@@ -383,6 +383,33 @@ Small quantities can be represented directly as decimals (e.g., 0.3), or using m
...
+ 2. Implementations MAY sanity-check keys before applying sysctl parameters. For example, only entries related to isolated subsystems (e.g. `net.*` on Linux) may be allowed.
```
Comment (euank) [inline spec/ace.md]: Perhaps we should have:

3. Implementations SHOULD only allow sysctl parameters that are safely namespaced and will not impact other pods.

**Comment by euank** [comment]
I tossed on some nits, but I think this makes sense, both in terms of naming and as an isolator.

File: spec/ace.md [hunk 5]
```diff
@@ -383,6 +383,33 @@ Small quantities can be represented directly as decimals (e.g., 0.3), or using m
 
 **NOTE**: Network limits MUST NOT apply to localhost communication between apps in a pod.
 
+### Environment Isolators
+
+An application or a group of applications may often require some tweaking to the environment where they run, in order to perform in an optimal way.
+Environment isolators take care of setting up the runtime environment as needed.
+
+#### environment/sysctl
```
Comment (jonboulle) [inline spec/ace.md]: Do we want to namespace this more specifically to Linux or *nix? /cc @mpasternacki if he has any thoughts from the BSD perspective

File: spec/ace.md (hunk 4 above)
Comment (jonboulle) [inline spec/ace.md]: maybe "For example, a Linux implementation might only allow entries related to a specific subtree like `net.*`"

File: spec/ace.md [hunk 6]
```diff
@@ -383,6 +383,33 @@ Small quantities can be represented directly as decimals (e.g., 0.3), or using m
...
+* dictionary of `sysctl(8)` key-value entries to set kernel runtime parameters.
+
+```json
+"name": "environment/sysctl",
+"value": {
+  "sysctl": {
+    "net.ipv4.tcp_mtu_probing": "1",
+    "net.ipv4.tcp_keepalive_time": "300",
+    "net.ipv4.tcp_keepalive_probes": "5",
+    "net.ipv4.tcp_keepalive_intvl": "15",
```
Comment (jonboulle) [inline spec/ace.md]: trailing ,

File: spec/ace.md (hunk 5 above)
Comment (lucab) [inline spec/ace.md]: I checked before venturing here, and at least FreeBSD has sysctl tuning support: https://www.freebsd.org/doc/handbook/configtuning-sysctl.html

Comment (jonboulle) [inline spec/ace.md]: Right, I'm just wondering if that suggests we could make this e.g. `os/unix/environment/sysctl`

File: spec/ace.md (hunk 4 above)
Comment (sttts) [inline spec/ace.md]: "will not impact other pods" is a very fuzzy term (compare https://github.com/kubernetes/kubernetes/pull/26057/files#diff-8dc23ab258695ee42154d4d1238c36efR323). Being namespaced in certain namespaces is much clearer defined (might differ by kernel version though).

Also note, that being namespaced of course depends on which namespaces are actually used (ipc, net, etc.) and not shared with the host. This should have influence on the allowed sysctls.

Comment (sttts) [inline spec/ace.md]: What about the error behavior after validation? Should a container refuse to launch?

File: spec/ace.md (hunk 1 above)
Comment (lucab) [inline spec/ace.md]: Ack.

File: spec/ace.md (hunk 2 above)
Comment (lucab) [inline spec/ace.md]: I agree this is missing some introduction text here. And thanks for the suggestion!

File: spec/ace.md (hunk 3 above)
Comment (lucab) [inline spec/ace.md]: Mostly because KISS and because merging behavior requires some kind of arbitrary priority ordering.

File: spec/ace.md (hunk 4 above)
Comment (lucab) [inline spec/ace.md]: Summing up a bit of discussion, it is desirable for this to be a SHOULD but without getting into specifying low-level stuff. Consumers are also interested in different behaviors after validation (apply anyway, ignore, fail-hard), but this is something that shouldn't be decided by spec, and left to runtime. As such, the middle-ground would be:
```
Implementations SHOULD validate keys before applying sysctl parameters. 
For example, a Linux implementation might only allow entries related to a 
specific namespaced subtree like `net.*`. In such case, implementations MUST
either ignore forbidden parameters or refuse to run the pod alltogether.
```

File: spec/ace.md (hunk 5 above)
Comment (lucab) [inline spec/ace.md]: I'm more or less neutral here and your scoping also makes sense (maybe dropping the `environment` part). Just for future proofing we may have a `posix-limits` and a `linux-limits` upcoming after this, how would you scope those too in you mind?

File: spec/ace.md [hunk 7]
```diff
@@ -383,6 +383,33 @@ Small quantities can be represented directly as decimals (e.g., 0.3), or using m
...
+* Scope: pod
+
+**Parameters:**
+
+* dictionary of `sysctl(8)` key-value entries to set kernel runtime parameters.
+
+```json
+"name": "environment/sysctl",
+"value": {
+  "sysctl": {
```
Comment (lucab) [inline spec/ace.md]: Heads-up: spurious label, this was not supposed to be here.

**Comment by jonboulle** [comment]
bump - this still needs a rework right?

**Comment by lucab** [comment]
It does, plus a decision on https://github.com/appc/spec/pull/647#discussion_r75478277

File: spec/ace.md (hunk 5 above)
Comment (jonboulle) [inline spec/ace.md]: @lucab I would have expected them to be `os/linux/limits` and `os/posix/limits` or similar, per the existing ones.

Comment (lucab) [inline spec/ace.md]: Ack, I guess we can settle on `os/unix/sysctl` - `os/unix/limits` - `os/linux/limits`. This is aligned with current naming trend.

**Comment by lucab** [comment]
Second iteration of this: renamed isolator to `os/unix/sysctl` and reworded the key validation note. PTAL.

**Comment by jonboulle** [comment]
LGTM
@vbatts any comments?

**Comment by euank** [comment]
LGTM as well

**Pull request merged**